File: content/LLD/Reference-Code/12. Factory Pattern/bad_example.py
from abc import ABC, abstractmethod


class Food(ABC):
    @abstractmethod
    def prepare(self):
        pass


class Pizza(Food):
    def prepare(self):
        print("Preparing pizza")


class Burger(Food):
    def prepare(self):
        print("Preparing burger")


# Responsible for making objects
class RestrauntService:
    def create_order(self, food_type: str):
        if food_type == "pizza":
            f = Pizza()
        elif food_type == "burger":
            f = Burger()
        else:
            print("Invalid food type")
            return None
        f.prepare()
        return f


restraunt_service = RestrauntService()
restraunt_service.create_order("pizza")
restraunt_service.create_order("burger")

File: content/LLD/Reference-Code/12. Factory Pattern/BadExample.java
public class BadExample {
    interface Food {
        void prepare();
    }

    static class Pizza implements Food {
        @Override
        public void prepare() {
            System.out.println("Preparing pizza");
        }
    }

    static class Burger implements Food {
        @Override
        public void prepare() {
            System.out.println("Preparing burger");
        }
    }

    // Responsible for making objects directly inside service
    static class RestaurantService {
        public Food createOrder(String foodType) {
            Food f;
            if ("pizza".equalsIgnoreCase(foodType)) {
                f = new Pizza();
            } else if ("burger".equalsIgnoreCase(foodType)) {
                f = new Burger();
            } else {
                System.out.println("Invalid food type");
                return null;
            }
            f.prepare();
            return f;
        }
    }

    public static void main(String[] args) {
        RestaurantService restaurantService = new RestaurantService();
        restaurantService.createOrder("pizza");
        restaurantService.createOrder("burger");
    }
}
